File: solutions/python/saddle-points/1/saddle_points.py
# With reverence and eternal respect to Shree DR.MDD

def rotate(board):
    return [[row[i] for row in board] for i in range(len(board[0]))]


def saddle_points(matrix):
    marks = []

    if not matrix:
        return marks
    if any(len(line) != len(matrix[0]) for line in matrix):
        raise ValueError("irregular matrix")

    transposed = rotate(matrix)

    for r in range(len(matrix)):
        for c in range(len(matrix[0])):
            val = matrix[r][c]
            if val == max(matrix[r]) and val == min(transposed[c]):
                marks.append({"row": r + 1, "column": c + 1})
    return marks

File: solutions/fortran/saddle-points/1/saddle_points.f90
! For my Shree DR.MDD
module saddle_points
   implicit none

   type :: point_t
      integer :: row
      integer :: column
   end type point_t

contains

   function saddlePoints(matrix) result(pairs)
      integer, intent(in) :: matrix(:, :)
      type(point_t), allocatable :: pairs(:)
      integer :: r, c, val
      integer, allocatable :: row_high(:), col_low(:)

      row_high = maxval(matrix, 2)
      col_low = minval(matrix, 1)
      allocate(pairs(0))

      do r = lbound(matrix, 1), ubound(matrix, 1)
         do c = lbound(matrix, 2), ubound(matrix, 2)
            val = matrix(r, c)
            if ((val == row_high(r)) .and. (val == col_low(c))) then
               pairs = [pairs, point_t(row=r, column=c)]
            end if
         end do
      end do

   end function saddlePoints

end module saddle_points
